v2/component/launcher.go

```diff
@@ -441,3 +511,181 @@ func (l *Launcher) RunComponent(ctx context.Context, cmd string, args ...string)
 
 	return l.metadataClient.PublishExecution(ctx, execution, outputParameters, outputArtifacts)
 }
+
+func getRuntimeArtifactSchema(rta *pipeline_spec.RuntimeArtifact) (string, error) {
+	switch t := rta.Type.Kind.(type) {
+	case *pipeline_spec.ArtifactTypeSchema_InstanceSchema:
+		return t.InstanceSchema, nil
+	case *pipeline_spec.ArtifactTypeSchema_SchemaTitle:
+		return "title: " + t.SchemaTitle, nil
+	case *pipeline_spec.ArtifactTypeSchema_SchemaUri:
+		return "", fmt.Errorf("SchemaUri is unsupported, found in RuntimeArtifact %+v", rta)
+	default:
+		return "", fmt.Errorf("unknown type %T in RuntimeArtifact %+v", t, rta)
+	}
+}
+
+func mergeRuntimeArtifacts(src, dst *pipeline_spec.RuntimeArtifact) {
+	if len(src.Uri) > 0 {
+		dst.Uri = src.Uri
+	}
+
+	if src.Metadata != nil {
+		if dst.Metadata == nil {
+			dst.Metadata = src.Metadata
+		} else {
+			for k, v := range src.Metadata.Fields {
+				dst.Metadata.Fields[k] = v
+			}
+		}
+	}
+}
+
+// TODO(neuromage): Move these helper functions to a storage package and add tests.
+func uploadFile(ctx context.Context, bucket *blob.Bucket, localFilePath, blobFilePath string) error {
+	errorF := func(err error) error {
+		return fmt.Errorf("uploadFile(): unable to complete copying %q to remote storage %q: %w", localFilePath, blobFilePath, err)
+	}
+
+	w, err := bucket.NewWriter(ctx, blobFilePath, nil)
+	if err != nil {
+		return errorF(fmt.Errorf("unable to open writer for bucket: %w", err))
+	}
+
+	r, err := os.Open(localFilePath)
+	if err != nil {
+		return errorF(fmt.Errorf("unable to open local file %q for reading: %w", localFilePath, err))
+	}
+	defer r.Close()
+
+	if _, err = io.Copy(w, r); err != nil {
+		return errorF(fmt.Errorf("unable to complete copying: %w", err))
+	}
+
+	if err = w.Close(); err != nil {
+		return errorF(fmt.Errorf("failed to close Writer for bucket: %w", err))
+	}
+
+	return nil
+}
+
+func downloadFile(ctx context.Context, bucket *blob.Bucket, blobFilePath, localFilePath string) error {
+	errorF := func(err error) error {
+		return fmt.Errorf("downloadFile(): unable to complete copying %q to local storage %q: %w", blobFilePath, localFilePath, err)
+	}
+
+	r, err := bucket.NewReader(ctx, blobFilePath, nil)
+	if err != nil {
+		return errorF(fmt.Errorf("unable to open reader for bucket: %w", err))
+	}
+	defer r.Close()
+
+	localDir := filepath.Dir(localFilePath)
+	if err := os.MkdirAll(localDir, 0644); err != nil {
+		return errorF(fmt.Errorf("failed to create local directory %q: %w", localDir, err))
```

Comment:
Because the above wrapping message already contains all the useful information for debug,
I think each return error statement no longer need to wrap with their specific messages.

The different messages in these error messages will very likely be duplicates of errors returned when calling `os.MkdirAll`, `bucket.NewReader`, `os.Create`, etc.

Reply:
The two error handling comments apply to other code too

Reply:
I think there's some value in knowing which of these operations failed though. I'd rather err on adding potentially redundant error messages for more information here for now while we test it out.

Reply:
Sounds like a good plan